_source/_posts/2021-08-04-microprofile-content-negotaion.adoc

```diff
@@ -0,0 +1,751 @@
+----
+
+[source,http]
+----
+HTTP/1.1 200 OK
+Content-Language: en
+Content-Type: application/json
+Vary: Accept-Language
+content-encoding: gzip
+content-length: 154
+
+{
+    "expression": {
+        "left": {
+            "numberOfDice": 2,
+            "numberOfFaces": 8
+        },
+        "operation": "ADD",
+        "right": {
+            "value": 1
+        }
+    },
+    "results": [
+    ...
+    ],
+    "value": 8
+}
+----
+====
+
+So far, so good!
+
+== Different Responses for Different Clients
+
+Up until now, the simple `/roll` endpoint has been returning JSON, which is excellent if you are building a client that accepts JSON, but it's not very user-friendly for text-based clients like HTTPie or curl. Two changes are needed to support a plain text response, annotate endpoint with an `@Produces` annotation and add a `MessageBodyWriter` that converts the dice parser's `ResultTree` object into text.
+
+Add the following `@Produces` annotation to `DiceResource`, to indicate it supports both `text/plain` and `application/json`:
+
+[source,java]
+----
+@Path("/roll")
+@Produces({MediaType.TEXT_PLAIN, MediaType.APPLICATION_JSON })
+public class DiceResource  {
+...
+----
+
+If you make a request to the `/roll` endpoint now, you will get an ugly response due to a call to the `toString()` method on `ResultTree`. 🤢
+
+To fix this, create a new class `TextDiceTreeMessageBodyWriter.java` that implements `MessageBodyWriter<ResultTree>`, which render something a little easier on the eyes:
+
+====
+[source,java]
+----
+package com.example;
+
+import dev.diceroll.parser.Dice;
+import dev.diceroll.parser.ResultTree;
+
+import javax.ws.rs.Produces;
+import javax.ws.rs.WebApplicationException;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.MultivaluedMap;
+import javax.ws.rs.ext.MessageBodyWriter;
+import javax.ws.rs.ext.Provider;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Type;
+import java.nio.charset.StandardCharsets;
+
+@Provider
+@Produces(MediaType.TEXT_PLAIN) // <.>
+public class TextDiceTreeMessageBodyWriter implements MessageBodyWriter<ResultTree> {
+
+    @Override
+    public boolean isWriteable(Class<?> type, Type genericType, Annotation[] annotations, MediaType mediaType) {
+        return type == ResultTree.class; // <.>
+    }
+
+    @Override
+    public void writeTo(ResultTree resultTree,
+                        Class<?> type, Type genericType,
+                        Annotation[] annotations,
+                        MediaType mediaType,
+                        MultivaluedMap<String, Object> httpHeaders,
+                        OutputStream entityStream) throws IOException, WebApplicationException {
+
+        String result = Dice.debug(resultTree); // <.>
+        entityStream.write(result.getBytes(StandardCharsets.UTF_8)); // <.>
+    }
+}
+----
+
+<.> Add a `Produces` annotation to the class to mark the `MediaType's it supports.
```

Comment:
```suggestion
<.> Add a `Produces` annotation to the class to mark the `MediaType`s it supports.
```
correct backtick punctuation

Reply:
I just bumped into this, the trailing `s` causes the code block to not render.  (accepting changes but also dropping the trailing `s`)